tensorflow_serving/tools/docker/Dockerfile.devel-mkl

```diff
@@ -0,0 +1,129 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+FROM ubuntu:16.04 as base_build
```

Comment:
Let's try this with 18.04, libpng12-dev below should become libpng-dev, and hopefully everything else should work.

Reply:
That looks like the server just isn’t up. Our unit tests are passing with 18.04 meaning that the server can be queried using the tests you’ve been seeing.

Reply:
I tested this inside container too. Manually launching `tensorflow-model-server` inside screen session. and querying from outside of the screen session. I got the same error. But the thing is, i ran same example as i did here for `16.04` and it worked fine without any issues. 🤔

Reply:
Yeah it was fine when i tested for `half_plus_two` model. It's failing with this pre-trained model.
https://github.com/tensorflow/models/tree/master/official/resnet#pre-trained-model. 
i tested with this http://download.tensorflow.org/models/official/20181001_resnet/savedmodels/resnet_v1_fp32_savedmodel_NCHW_jpg.tar.gz

Reply:
```
docker run --name=tfserving_rest_mkl --rm -d -p 8501:8501 -v "/home/nfsshare/karthik/git-repo/oob_work/tfserving_bm/client/mkl/resnet_v1_fp32_savedmodel_NCHW_jpg:/models/resnet" -e MODEL_NAME=resnet -e OMP_NUM_THREADS=36 -e MKLDNN_VERBOSE=1 -e TENSORFLOW_SESSION_PARALLELISM=9 tensorflow/serving:latest-mkl
565ac6aa94c442ad45c4a37a75e8b7f3c43e9fb03b326da89f1efbaf8883e597
```
```
docker ps | grep tfserving
565ac6aa94c4        tensorflow/serving:latest-mkl   "/usr/bin/tf_serving…"   About a minute ago   Up About a minute   8500/tcp, 0.0.0.0:8501->8501/tcp   tfserving_rest_mkl
```
```
$ docker logs 565ac6aa94c4
2018-12-06 05:47:45.163877: I tensorflow_serving/model_servers/server.cc:82] Building single TensorFlow model file config:  model_name: resnet model_base_path: /models/resnet
2018-12-06 05:47:45.164104: I tensorflow_serving/model_servers/server_core.cc:461] Adding/updating models.
2018-12-06 05:47:45.164124: I tensorflow_serving/model_servers/server_core.cc:558]  (Re-)adding model: resnet
2018-12-06 05:47:45.265461: I tensorflow_serving/core/basic_manager.cc:739] Successfully reserved resources to load servable {name: resnet version: 1538686758}
2018-12-06 05:47:45.265503: I tensorflow_serving/core/loader_harness.cc:66] Approving load for servable version {name: resnet version: 1538686758}
2018-12-06 05:47:45.265514: I tensorflow_serving/core/loader_harness.cc:74] Loading servable version {name: resnet version: 1538686758}
2018-12-06 05:47:45.265544: I external/org_tensorflow/tensorflow/contrib/session_bundle/bundle_shim.cc:363] Attempting to load native SavedModelBundle in bundle-shim from: /models/resnet/1538686758
2018-12-06 05:47:45.265563: I external/org_tensorflow/tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: /models/resnet/1538686758
2018-12-06 05:47:45.273941: I external/org_tensorflow/tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { serve }

User settings:

   KMP_AFFINITY=granularity=fine,verbose,compact,1,0
   KMP_BLOCKTIME=1
   KMP_SETTINGS=1
   OMP_NUM_THREADS=36

Effective settings:

   KMP_ABORT_DELAY=0
   KMP_ADAPTIVE_LOCK_PROPS='1,1024'
   KMP_ALIGN_ALLOC=64
   KMP_ALL_THREADPRIVATE=288
   KMP_ATOMIC_MODE=2
   KMP_BLOCKTIME=1
   KMP_CPUINFO_FILE: value is not defined
   KMP_DETERMINISTIC_REDUCTION=false
   KMP_DEVICE_THREAD_LIMIT=2147483647
   KMP_DISP_HAND_THREAD=false
   KMP_DISP_NUM_BUFFERS=7
   KMP_DUPLICATE_LIB_OK=false
   KMP_FORCE_REDUCTION: value is not defined
   KMP_FOREIGN_THREADS_THREADPRIVATE=true
   KMP_FORKJOIN_BARRIER='2,2'
   KMP_FORKJOIN_BARRIER_PATTERN='hyper,hyper'
   KMP_FORKJOIN_FRAMES=true
   KMP_FORKJOIN_FRAMES_MODE=3
   KMP_GTID_MODE=3
   KMP_HANDLE_SIGNALS=false
   KMP_HOT_TEAMS_MAX_LEVEL=1
   KMP_HOT_TEAMS_MODE=0
   KMP_INIT_AT_FORK=true
   KMP_INIT_WAIT=2048
   KMP_ITT_PREPARE_DELAY=0
   KMP_LIBRARY=throughput
   KMP_LOCK_KIND=queuing
   KMP_MALLOC_POOL_INCR=1M
   KMP_NEXT_WAIT=1024
   KMP_NUM_LOCKS_IN_BLOCK=1
   KMP_PLAIN_BARRIER='2,2'
   KMP_PLAIN_BARRIER_PATTERN='hyper,hyper'
   KMP_REDUCTION_BARRIER='1,1'
   KMP_REDUCTION_BARRIER_PATTERN='hyper,hyper'
   KMP_SCHEDULE='static,balanced;guided,iterative'
   KMP_SETTINGS=true
   KMP_SPIN_BACKOFF_PARAMS='4096,100'
   KMP_STACKOFFSET=64
   KMP_STACKPAD=0
   KMP_STACKSIZE=4M
   KMP_STORAGE_MAP=false
   KMP_TASKING=2
   KMP_TASKLOOP_MIN_TASKS=0
   KMP_TASK_STEALING_CONSTRAINT=1
   KMP_TEAMS_THREAD_LIMIT=72
   KMP_TOPOLOGY_METHOD=all
   KMP_USER_LEVEL_MWAIT=false
   KMP_VERSION=false
   KMP_WARNINGS=true
   OMP_AFFINITY_FORMAT='OMP: pid %P tid %T thread %n bound to OS proc set {%a}'
   OMP_ALLOCATOR=omp_default_mem_alloc
   OMP_CANCELLATION=false
   OMP_DEFAULT_DEVICE=0
   OMP_DISPLAY_AFFINITY=false
   OMP_DISPLAY_ENV=false
   OMP_DYNAMIC=false
   OMP_MAX_ACTIVE_LEVELS=2147483647
   OMP_MAX_TASK_PRIORITY=0
   OMP_NESTED=false
   OMP_NUM_THREADS='36'
   OMP_PLACES: value is not defined
   OMP_PROC_BIND='intel'
   OMP_SCHEDULE='static'
   OMP_STACKSIZE=4M
   OMP_TARGET_OFFLOAD=DEFAULT
   OMP_THREAD_LIMIT=2147483647
   OMP_TOOL=enabled
   OMP_TOOL_LIBRARIES: value is not defined
   OMP_WAIT_POLICY=PASSIVE
   KMP_AFFINITY='verbose,warnings,respect,granularity=fine,compact,1,0'

2018-12-06 05:47:45.316887: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:182] Restoring SavedModel bundle.
2018-12-06 05:47:45.718556: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:132] Running initialization op on SavedModel bundle.
2018-12-06 05:47:45.739747: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:285] SavedModel load for tags { serve }; Status: success. Took 474162 microseconds.
2018-12-06 05:47:45.739808: I tensorflow_serving/servables/tensorflow/saved_model_warmup.cc:83] No warmup data file found at /models/resnet/1538686758/assets.extra/tf_serving_warmup_requests
2018-12-06 05:47:45.740221: I tensorflow_serving/core/loader_harness.cc:86] Successfully loaded servable version {name: resnet version: 1538686758}
2018-12-06 05:47:45.774851: I tensorflow_serving/model_servers/server.cc:286] Running gRPC ModelServer at 0.0.0.0:8500 ...
[warn] getaddrinfo: address family for nodename not supported
2018-12-06 05:47:45.788121: I tensorflow_serving/model_servers/server.cc:302] Exporting HTTP/REST API at:localhost:8501 ...
[evhttp_server.cc : 237] RAW: Entering the event loop ...
```
and when i query. it throws this message in terminal as shown for `REST`. and container is completely wiped out..i couldn't see any traces even with `docker ps -a`.

Reply:
Hmm if it works with half plus two, my guess is that something is unhappy with the mkl optimizations causing a crash. The —rm option on docker is maybe cleaning up the image after the crash.

Reply:
yeah, my other colleagues are testing mkl optimizations support in `18.04`. So what do you think 🤔 , should we get this IN first and then will create a separate PR once we have fixes for `18.04` upgrade.

Reply:
Sure, that’s fine :)

Reply:
Thank you 👍

Reply:
@gautamvasudevan I could to the point loading model which uses `mkl` stuff. But with `bazel test` i see time out issues
```Waiting for HTTP endpoint
2018-12-07 00:33:05.272104: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:132] Running initialization op on SavedModel bundle.
2018-12-07 00:33:05.295142: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:285] SavedModel load for tags { serve }; Status: success. Took 506975 microseconds.
2018-12-07 00:33:05.295186: I tensorflow_serving/servables/tensorflow/saved_model_warmup.cc:83] No warmup data file found at /models/saved_model_resnet_v1_fp32_savedmodel_NCHW_jpg/1538686758/assets.extra/tf_serving_warmup_requests
2018-12-07 00:33:05.295258: I tensorflow_serving/core/loader_harness.cc:86] Successfully loaded servable version {name: saved_model_resnet_v1_fp32_savedmodel_NCHW_jpg version: 1538686758}
2018-12-07 00:33:05.493795: I tensorflow_serving/model_servers/server.cc:286] Running gRPC ModelServer at 0.0.0.0:8500 ...
[warn] getaddrinfo: address family for nodename not supported
2018-12-07 00:33:05.512588: I tensorflow_serving/model_servers/server.cc:302] Exporting HTTP/REST API at:localhost:8501 ...
[evhttp_server.cc : 237] RAW: Entering the event loop ...
Polling for HTTP endpoint for image tensorflow/serving:latest-mkl...
Will query ModelServer for 3 (max) attempts...
Querying TF ModelServer (attempt: 1)
current_directory: /tmp/bazel/output/execroot/tf_serving/bazel-out/k8-opt/bin/tensorflow_serving/tools/docker/tests/unittest_dockerfile_mkl.runfiles/tf_serving
Terminated

TIMEOUT: //tensorflow_serving/tools/docker/tests:unittest_dockerfile_mkl (Summary)
      /tmp/bazel/output/execroot/tf_serving/bazel-out/k8-opt/testlogs/tensorflow_serving/tools/docker/tests/unittest_dockerfile_mkl/test.log
Target //tensorflow_serving/tools/docker/tests:unittest_dockerfile_mkl up-to-date:
  bazel-bin/tensorflow_serving/tools/docker/tests/unittest_dockerfile_mkl
INFO: Elapsed time: 75.454s, Critical Path: 75.11s
INFO: 1 process: 1 local.
INFO: Build completed, 1 test FAILED, 2 total actions
//tensorflow_serving/tools/docker/tests:unittest_dockerfile_mkl         TIMEOUT in 75.0s
  /tmp/bazel/output/execroot/tf_serving/bazel-out/k8-opt/testlogs/tensorflow_serving/tools/docker/tests/unittest_dockerfile_mkl/test.log

INFO: Build completed, 1 test FAILED, 2 total actions```
i have sent request individually after the test failure. I could see the result as below.
```python ../../../../../../../../../tensorflow_serving/example/resnet_client.py saved_model_resnet_v1_fp32_savedmodel_NCHW_jpg
Prediction class: 286, avg latency: 65.7438 ms```
Not sure what's wrong with `bazel test`.

Reply:
75 seconds to run the test is longer than a small test now. See [here](https://docs.bazel.build/versions/master/test-encyclopedia.html#role-of-the-test-runner) for details on how bazel works with test timeouts. Specifically this will have to be a medium size test now (with a 300s timeout).

Reply:
yeah that helped to see what causing the issue. Here is the culprit.
```Traceback (most recent call last):
  File "../../../../../../../../../tensorflow_serving/example/resnet_client.py", line 77, in <module>
    main()
  File "../../../../../../../../../tensorflow_serving/example/resnet_client.py", line 51, in main
    dl_request = requests.get(IMAGE_URL, stream=True)
  File "/usr/local/lib/python2.7/dist-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='tensorflow.org', port=443): Max retries exceeded with url: /images/blogs/serving/cat.jpg (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f97893d3b10>: Failed to establish a new connection: [Errno 110] Connection timed out',))```
https://github.com/tensorflow/serving/blob/master/tensorflow_serving/example/resnet_client.py#L47 . I'm running this behind proxy. I tried exporting my proxy `http` and `https` too.. that didn't help. Do we have a different way to pass for `bazel test`

Reply:
Would it be acceptable to include the cat.jpg within the container rather than always grabbing it from the url?

Reply:
oh Tony, `bazel test` are running outside of container. Basically it spins up container based on image tag provided and sends predict request to that container.

Reply:
Correct. It validates the container as it would be used, not within the container.

That said, the reason we used half_plus_two for validation was that it made the tests simple and hermetic, without introducing too much in the way of input processing on the client (zero, in fact). It would be ideal to provide a model that similarly took in simple data and performed whatever minimal computation was necessary to invoke MKL.